### PR TITLE
add setup and kill telemetry events - use nodejs runtime in telemetry api route

### PR DESCRIPTION
**Pull request by eliobricenov**

### CODE REVIEW

File: packages/core/src/telemetry/service.ts [hunk 1]
```diff
@@ -143,7 +143,7 @@ export class TelemetryService {
     fs.writeFileSync(telemetryEventsFilePath, serializedEvents);
 
     child_process.spawn(process.execPath, [
-      path.join(__dirname, "flush-detached.js"),
+      path.join(__dirname, "detached-flush.js"),
```
Comment (eliobricenov) [inline packages/core/src/telemetry/service.ts:146]: 🙃

File: packages/core/src/telemetry/service.ts [hunk 2]
```diff
@@ -15,7 +15,7 @@ import { getGitRemoteUrl } from "@/telemetry/remote";
 
 type TelemetryEvent = {
   event: string;
-  payload: object;
+  payload?: object;
```
Comment (eliobricenov) [inline packages/core/src/telemetry/service.ts:18]: I don't think we need to enforce a payload, only event is really necessary.

File: packages/core/src/telemetry/detached-flush.js [hunk 3]
```diff
@@ -19,7 +19,7 @@ async function detachedFlush() {
     events.map(async (event) => {
       await fetch(telemetryUrl, {
         method: "POST",
-        body: JSON.stringify(event.payload),
+        body: JSON.stringify(event),
```
Comment (eliobricenov) [inline packages/core/src/telemetry/detached-flush.js:22]: We want to include the whole event object, including event name and meta.

File: docs/pages/api/telemetry/index.ts [hunk 4]
```diff
@@ -1,44 +1,32 @@
...
-    { status: 200 }
-  );
+  res.status(200).json({
+    message: "Telemetry data processed successfully.",
+  });
 
   await analytics.track({
-    ...data,
+    ...req.body,
   });
```
Comment (0xOlias) [inline docs/pages/api/telemetry/index.ts:31]: Is there a reason you need to copy/spread here?

File: docs/pages/api/telemetry/index.ts [hunk 5]
```diff
@@ -1,44 +1,32 @@
...
-      { status: 405 }
-    );
+    return res.status(405).json({
+      error: `Method ${req.method} Not Allowed`,
+    });
   }
 
   const analytics = new Analytics({
     writeKey: process.env.SEGMENT_WRITE_KEY,
   });
```
Comment (0xOlias) [inline docs/pages/api/telemetry/index.ts:23]: I wonder if, now that this is an API route, you can initialize this outside the function scope.

File: packages/core/src/telemetry/detached-flush.js (hunk 3 above)
Comment (0xOlias) [reply packages/core/src/telemetry/detached-flush.js:22]: Makes sense, think this was an oversight on my part!

File: packages/core/src/telemetry/detached-flush.js [hunk 6]
```diff
@@ -19,13 +19,15 @@ async function detachedFlush() {
...
         headers: {
           "Content-Type": "application/json",
         },
       });
     })
-  );
+  ).finally(() => {
+    fs.rmSync(eventsFilePath);
+  });
```
Comment (0xOlias) [inline packages/core/src/telemetry/detached-flush.js:30]: I would probably change this to
```ts
try {
  await Promise.all( ... )
} catch (error) {
  console.error(error)
}

fs.rmSync(eventsFilePath)
```

[See this](https://stackoverflow.com/a/58056140).

File: packages/core/src/telemetry/detached-flush.test.ts [hunk 7]
```diff
@@ -27,7 +27,10 @@ test("detached flush script should run without error", async ({ common }) => {
     );
   });
 
-  rmSync(telemetryEventsFilePath);
+  // if the script ran successfully, the temporary file should be deleted
+  if (existsSync(telemetryEventsFilePath)) {
+    rmSync(telemetryEventsFilePath);
+  }
```
Comment (0xOlias) [inline packages/core/src/telemetry/detached-flush.test.ts:33]: I think this makes more sense as an assertion

```ts
expect(existsSync(telemetryEventsFilePath)).toBeFalsy()
```

Or something like that. If the script is leaving around dead files, that should be a test failure.

File: packages/core/src/telemetry/service.ts (hunk 1 above)
Comment (0xOlias) [reply packages/core/src/telemetry/service.ts:146]: 3a7

**Review by 0xOlias — COMMENTED** [review]
Looks great.

Had a thought - I wonder if it makes sense to use `os.tmpdir()` rather than `/.ponder/` for the location of the `telemetry-events.json` file.

File: packages/core/src/telemetry/detached-flush.test.ts (hunk 7 above)
Comment (eliobricenov) [reply packages/core/src/telemetry/detached-flush.test.ts:33]: great call

File: docs/pages/api/telemetry/index.ts (hunk 4 above)
Comment (eliobricenov) [reply docs/pages/api/telemetry/index.ts:31]: not at all - I missed changing this when we no longer needed to do `const data = await req.json()` from the edge version

File: docs/pages/api/telemetry/index.ts (hunk 5 above)
Comment (eliobricenov) [reply docs/pages/api/telemetry/index.ts:23]: we can!

File: packages/core/src/telemetry/detached-flush.js (hunk 6 above)
Comment (0xOlias) [reply packages/core/src/telemetry/detached-flush.js]: Hmm - I think it's fine to ignore any errors coming from the script tbh. The only real source of errors is segment HTTP request failure/error. Our test for the script is pretty thorough.

**Pull request merged**